unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp

```diff
@@ -0,0 +1,340 @@
+/*******************************************************************\
+
+ Module: Unit tests for parsing generic classes
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <testing-utils/catch.hpp>
+
+#include <util/config.h>
+#include <util/cmdline.h>
+#include <util/language.h>
+#include <util/prefix.h>
+
+#include <java_bytecode/java_bytecode_language.h>
+
+#include <iostream>
+#include <testing-utils/load_java_class.h>
+#include <testing-utils/require_type.h>
+
+SCENARIO(
+  "java_bytecode_parse_functions_with_generics",
+  "[core][java_bytecode][java_bytecode_parse_generics]")
+{
+  const symbol_tablet &new_symbol_table =
+    load_java_class(
+      "FunctionsWithGenerics", "./java_bytecode/java_bytecode_parse_generics");
+
+  std::string class_prefix = "java::FunctionsWithGenerics";
+
+  WHEN("Parsing processReturnSame")
+  {
+    THEN("There should be a symbol for the function")
+    {
+      const std::string func_name = ".processReturnSame";
+      const std::string func_descriptor = ":(LGeneric;)LGeneric;";
+      const std::string process_func_name =
+        class_prefix + func_name + func_descriptor;
+
+      REQUIRE(new_symbol_table.has_symbol(process_func_name));
+
+      const symbolt func_symbol =
+        new_symbol_table.lookup_ref(process_func_name);
+      const code_typet func_code = to_code_type(func_symbol.type);
+
+      THEN("It has parameter x pointing to Generic")
+      {
+        code_typet::parametert param_x =
+          require_type::require_parameter(func_code, "x");
+        require_type::require_pointer(
+          param_x.type(), symbol_typet("java::Generic"));
+
+        THEN("x is generic with parameter pointing to java.lang.Integer")
+        {
+          REQUIRE(is_java_generic_type(param_x.type()));
+
+          const java_generic_typet generic_x =
+            to_java_generic_type(param_x.type());
```

Comment:
For this pattern of `is_java_generic_type`, followed by `to_java_generic_type`, it's probably worth factoring out into a `require_java_generic_type` helper function - https://github.com/diffblue/cbmc/pull/1460/files#diff-6dc35e009e58da05abb1a1765acdd9f7R77 might have something you can steal.

Reply:
These look useful Chris - would you mind cherry picking that commit @svorenova (or alternatively you can pull it into a PR @chrisr-diffblue and then base this on top of it).